moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,1254 @@
+{
+// Constructor for the main ROS interface node
+JogROSInterface::JogROSInterface()
+{
+  ros::NodeHandle n;
+
+  // Read ROS parameters, typically from YAML file
+  if (!readParameters(n))
+    exit(EXIT_FAILURE);
+
+  // Load the robot model. This is needed by the worker threads.
+  model_loader_ptr_ = std::unique_ptr<robot_model_loader::RobotModelLoader>(new robot_model_loader::RobotModelLoader);
+
+  // Crunch the numbers in this thread
+  pthread_t joggingThread;
+  int rc = pthread_create(&joggingThread, nullptr, jog_arm::JogROSInterface::jogCalcThread, this);
+  if (rc)
+  {
+    ROS_FATAL_STREAM_NAMED(NODE_NAME, "Creating jog calculation thread failed");
+    return;
+  }
+
+  // Check collisions in this thread
+  pthread_t collisionThread;
+  rc = pthread_create(&collisionThread, nullptr, jog_arm::JogROSInterface::collisionCheckThread, this);
+  if (rc)
+  {
+    ROS_FATAL_STREAM_NAMED(NODE_NAME, "Creating collision check thread failed");
+    return;
+  }
+
+  // ROS subscriptions. Share the data with the worker threads
+  ros::Subscriber cmd_sub =
+      n.subscribe(ros_parameters_.cartesian_command_in_topic, 1, &JogROSInterface::deltaCartesianCmdCB, this);
+  ros::Subscriber joints_sub = n.subscribe(ros_parameters_.joint_topic, 1, &JogROSInterface::jointsCB, this);
+  ros::Subscriber joint_jog_cmd_sub =
+      n.subscribe(ros_parameters_.joint_command_in_topic, 1, &JogROSInterface::deltaJointCmdCB, this);
+  ros::topic::waitForMessage<sensor_msgs::JointState>(ros_parameters_.joint_topic);
+  ros::topic::waitForMessage<geometry_msgs::TwistStamped>(ros_parameters_.cartesian_command_in_topic);
+
+  // Publish freshly-calculated joints to the robot
+  // Put the outgoing msg in the right format (trajectory_msgs/JointTrajectory
+  // or std_msgs/Float64MultiArray).
+  ros::Publisher outgoing_cmd_pub;
+  if (ros_parameters_.command_out_type == "trajectory_msgs/JointTrajectory")
+    outgoing_cmd_pub = n.advertise<trajectory_msgs::JointTrajectory>(ros_parameters_.command_out_topic, 1);
+  else if (ros_parameters_.command_out_type == "std_msgs/Float64MultiArray")
+    outgoing_cmd_pub = n.advertise<std_msgs::Float64MultiArray>(ros_parameters_.command_out_topic, 1);
+
+  // Wait for low pass filters to stabilize
+  ros::Duration(10 * ros_parameters_.publish_period).sleep();
+
+  ros::Rate main_rate(1. / ros_parameters_.publish_period);
+
+  while (ros::ok())
+  {
+    ros::spinOnce();
+
+    pthread_mutex_lock(&shared_variables_.new_traj_mutex);
+    trajectory_msgs::JointTrajectory new_traj = shared_variables_.new_traj;
+    pthread_mutex_unlock(&shared_variables_.new_traj_mutex);
+
+    // Check for stale cmds
+    pthread_mutex_lock(&shared_variables_.incoming_cmd_stamp_mutex);
+    if ((ros::Time::now() - shared_variables_.incoming_cmd_stamp) <
+        ros::Duration(ros_parameters_.incoming_command_timeout))
+    {
+      // Mark that incoming commands are not stale
+      pthread_mutex_lock(&shared_variables_.command_is_stale_mutex);
```

Comment:
this lock can be moved outside the if statement since its used in both the if and else